### PR TITLE
Refactoring: threshold methods with histogram parameters

### PR DESCRIPTION
**Pull request by haesleinhuepf**
added method threshold_otsu_from_histogram as shortcut in case the histogram is known in advance

## Description
Hi all, hi @jni ,

this PR-draft refactors thresholding_otsu method so that one can call it also with a histogram instead of an image. I would like to use it in a [GPU-accelerated image processing library](https://github.com/clEsperanto/pyclesperanto_prototype) where we can determine the histogram on the GPU, but lack threshold-method implementations.

I would wonder what you think about this implementation. If it makes sense, I would offer to implement the same strategy for other threshold methods.

Best,
Robert

## Checklist

[x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
[ ] Gallery example in `./doc/examples` (new features only)
[ ] Benchmark in `./benchmarks`, if your changes aren't covered by an existing benchmark
[ ] Unit tests
[ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by emmanuelle** [comment]
Hi @haesleinhuepf thanks for starting this discussion! I'll let other devs chime in, but my gut feeling is that rather than adding one function to the API we could just add one keyword argument `histogram` (which would default to None) to the existing function. I see the interest of precomputing the histogram also for comparing different thresholding functions which require the histogram. 

@sciunto you've worked extensively on thresholding functions what do you think of these additions?

**Comment by jni** [comment]
I think the same as @emmanuelle: let's add `histogram=None` to the functions that benefit from a pre-computed histogram. We could potentially then also make `image=None`, and make sure one of the two is provided, with preference given to the histogram.

**Comment by haesleinhuepf** [comment]
Hey @emmanuelle and @jni ,

at the beginning I thought so, too. But then I saw that we need the histogram and the bin_centers array. Otherwise, we can only return the bin_index  and not the corresponding grey value. I didn't know how force the user to provide both together or both none. Assert statements came to my mind. Any better idea how to achieve this? 

Thanks for the feedback! 😃

**Comment by jni** [comment]
A great point that I had overlooked! 😅 

My preferred option given this would be to provide a tuple of `counts, bin_centers`, same as returned by `exposure.histogram`. Perhaps we also support a single array and then use the bin index (`bin_centers = np.arange(len(counts))`), which is convenient for uint16 and uint8 images.

**Comment by haesleinhuepf** [comment]
> Perhaps we also support a single array and then use the bin index (`bin_centers = np.arange(len(counts))`), which is convenient for uint16 and uint8 images.

Just to make sure I understand correctly: That would mean the threshold_otsu method returns a conceptually different thing (histogram index versus grey value) depending on input parameters. If that's ok, I'm happy to implement it.

**Comment by sciunto** [comment]
I'm +1 with you @emmanuelle and this can be applied to several of our functions as well.

**Comment by jni** [comment]
> a conceptually different thing 

It's not, the point is that for integer-valued images they are one and the same thing! So it is a shortcut for users of those images. If someone decides to misuse the shortcut, that's on them! 😜

At any rate, you can implement the tuple input version first and later we can decide if we want to provide this shortcut. 😊

**Comment by haesleinhuepf** [comment]
Alright @emmanuelle  and @jni ,

I just implemented it like you suggested. The function header is now
```
def threshold_otsu(image=None, nbins=256, hist=None, bin_centers=None)
```

If prefers the hist over the image as @jni suggested and returns the bin index in case no bin_centers are given. The only thing I didn't manage is the tuple-thingy @jni suggested. I assume one needs to add some more brakets to the  function defintion, but I'm not sure how this works with bin_centers being optional. You know, I just recently arrived in snake county. ;-)

Thanks again for your feedback. I appreciate it :-)

Let me know what you think!

Cheers,
Robert

**Comment by jni** [comment]
@haesleinhuepf I think you were expecting more magic than Python provides here. 😂  I merely meant that a single argument, `histogram=`, takes in as input a tuple of arrays, which you unpack inside the function:

```python
def threshold_otsu(image=None, nbins=256, *, histogram=None):
    if histogram is not None:
        counts, bin_centers = histogram
    else:
        counts, bin_centers = exposure.histogram(image, nbins)
```

or 

```python
def threshold_otsu(image=None, nbins=256, *, histogram=None):
    if histogram is not None:
        if isinstance(histogram, (tuple, list)):
            counts, bin_centers = histogram
        else:
            counts = histogram
            bin_centers = np.arange(counts.size)
    else:
        counts, bin_centers = exposure.histogram(image, nbins)
```

File: skimage/filters/thresholding.py
Comment (emmanuelle) [inline skimage/filters/thresholding.py:238]: Please find another name here as it conflicts with the `histogram` function imported from `exposure` at the top of the file (another solution would be to use it as `exposure.histogram` in the whole file, but it's probably better to choose another name, maybe just `hist`?)

Comment (haesleinhuepf) [reply skimage/filters/thresholding.py:238]: Good catch! It was called `hist` earlier. So we can stick to that shorter name.

**Comment by haesleinhuepf** [comment]
Hey @emmanuelle and @jni ,

this should ne it. Please let me know if there is anything else that should be adapted. If you like it, I'd be happy to do the same change for the other thresholding methods :-)

File: skimage/filters/thresholding.py
Comment (jni) [inline skimage/filters/thresholding.py:239]:
```suggestion
    """Return threshold value based on Otsu's method.

```

File: skimage/filters/thresholding.py
Comment (jni) [inline skimage/filters/thresholding.py:277]:
```suggestion
        raise Exception("Either image or hist must be provided.")
```

File: skimage/filters/thresholding.py
Comment (jni) [inline skimage/filters/thresholding.py:314]:
```suggestion
    threshold = bin_centers[idx]
```

I know it's not your code but might as well do some cleanup! This should clearly not affect the output!

File: skimage/filters/thresholding.py
Comment (jni) [inline skimage/filters/thresholding.py:250]:
```suggestion
    hist : array, or 2-tuple of arrays, optional
```

**Review by jni — APPROVED** [review]
Thank you @haesleinhuepf! For me this is ready! I'm excited to have this functionality throughout the module!

I've made some very minor formatting wording suggestions — please apply at your leisure. :blush:

**Comment by jni** [comment]
@scikit-image/core anyone else want to push the green button on this one? 😊

**Comment by haesleinhuepf** [comment]
> @scikit-image/core anyone else want to push the green button on this one? 😊

before you push any button: May I also adapt the other thresholding methods within this PR?

**Comment by jni** [comment]
@haesleinhuepf as you wish, but I am very much a "bird in the hand" kinda guy. 😂

**Comment by jni** [comment]
Actually it occurs to me that we should probably add a test for this code! 😂

**Comment by jni** [comment]
@haesleinhuepf would you mind adding that test? It should be the same as the test for otsu, but computing the histogram outside of the function call.

**Comment by haesleinhuepf** [comment]
Sure! Will add it/them [here](https://github.com/scikit-image/scikit-image/blob/master/skimage/filters/tests/test_thresholding.py#L222)

**Comment by jni** [comment]
@haesleinhuepf I've refactored the copy-pasted code into its own function, to avoid repeating code too much. =)

FYI you should generally not make PRs from master, but rather from a specific branch. This is because it is now impossible for you to have your master branch match ours after this PR merges. It's not a huge deal, just a bunch of branch renaming, but it's simpler to just start from a separate branch to begin with. =)

@scikit-image/core this is ready to merge imho, if anyone has some review cycles available!

File: skimage/filters/thresholding.py
Comment (mkcor) [inline skimage/filters/thresholding.py:258]:
```suggestion
        Each element is the number of pixels falling in each intensity bin.
```

File: skimage/filters/thresholding.py
Comment (mkcor) [inline skimage/filters/thresholding.py:260]:
```suggestion
        Each element is the value corresponding to the center of each intensity bin.
```

File: skimage/filters/thresholding.py
Comment (mkcor) [inline skimage/filters/thresholding.py:285]:
```suggestion
    Either image or hist must be provided. If hist is provided, the actual
```
I find it more direct this way.

File: skimage/filters/thresholding.py
Comment (mkcor) [inline skimage/filters/thresholding.py:257]: :thinking: Counts are integers... I can see you are casting them into floats line 279. But the `counts` returned by `exposure.histogram` have dtype `int`...

File: skimage/filters/thresholding.py
Comment (mkcor) [inline skimage/filters/thresholding.py:296]:
```suggestion
        Histogram to determine the threshold from, and possibly a corresponding array
```

File: skimage/filters/thresholding.py
Comment (mkcor) [inline skimage/filters/thresholding.py:366]: I would make the same suggestion as above. Also, I wonder if 'Alternative' is clear enough (it seems like we're referring to what was just said). I would rephrase along the lines of: "An alternative usage of this threshold function is to pass it only a histogram, not an image."

**Review by mkcor — COMMENTED** [review]
@jni I left a few comments, but I'm not far from approving!
PS: I have taken the liberty of committing the tiny cleanup suggestion.

File: skimage/filters/thresholding.py
Comment (jni) [reply skimage/filters/thresholding.py:257]: @mkcor yes, but basically every calling function was then converting them to float, so I thought that conversion might as well get rolled into here.

File: skimage/filters/thresholding.py
Comment (mkcor) [inline skimage/filters/thresholding.py:278]: Are the above two lines extra-indented, or is it just GitHub's rendering?

File: skimage/filters/thresholding.py
Comment (mkcor) [reply skimage/filters/thresholding.py:257]: Ok!

File: skimage/filters/thresholding.py
Comment (jni) [reply skimage/filters/thresholding.py:278]: this was my editor's indentation choice and I liked it! And it's PEP8 compliant. ;)

Comment (jni) [reply skimage/filters/thresholding.py:278]: (Did I already mention I hate black? =P)

File: skimage/filters/thresholding.py
Comment (jni) [reply skimage/filters/thresholding.py:296]: Even better I think:

```suggestion
        Histogram from which to determine the threshold from, and optionally a corresponding array
```

But this might need reflowing for PEP8.

Comment (mkcor) [reply skimage/filters/thresholding.py:296]: Agreed! Oops, almost: remove second 'from' (:

Comment (jni) [reply skimage/filters/thresholding.py:296]: Oops indeed! Edited. :facepalm:

File: skimage/filters/thresholding.py
Comment (mkcor) [reply skimage/filters/thresholding.py:278]: > this was my editor's indentation choice and I liked it! And it's PEP8 compliant. ;)

No, it's not! :confounded: See https://github.com/scikit-image/scikit-image/pull/5006#issuecomment-703691052:

* In the file skimage/filters/thresholding.py:

    Line 277:17: E126 continuation line over-indented for hanging indent
    Line 278:13: E121 continuation line under-indented for hanging indent

Comment (mkcor) [reply skimage/filters/thresholding.py:278]: > (Did I already mention I hate black? =P)

This I know! :rofl: I was surprised to see double quotes [here](https://github.com/scikit-image/scikit-image/pull/5006/files#diff-4f1fba6d7478009483dadb139f489db58130185867229af13354ae80a1b2ca3aR267). :wink:

File: skimage/filters/thresholding.py
Comment (mkcor) [inline skimage/filters/thresholding.py:368]:
```suggestion
        Histogram from which to determine the threshold, and optionally a
        corresponding array of bin center intensities.
        An alternative use of this function is to pass it only hist.
```

File: skimage/filters/thresholding.py
Comment (mkcor) [inline skimage/filters/thresholding.py:298]:
```suggestion
        Histogram from which to determine the threshold, and optionally a
        corresponding array of bin center intensities.
        An alternative use of this function is to pass it only hist.
```

**Review by mkcor — COMMENTED** [review]
@haesleinhuepf please note, in case you were going to commit on your branch again, that I have since committed a suggestion on it. Before resuming work locally, you would therefore need to sync up your branch with a `git pull`.

File: skimage/filters/thresholding.py
Comment (mkcor) [reply skimage/filters/thresholding.py:298]: @jni PEP8 ^^

File: skimage/filters/thresholding.py
Comment (haesleinhuepf) [reply skimage/filters/thresholding.py:278]: Maybe these double quote come from that Java developer who sent this PR ;-)

File: skimage/filters/thresholding.py
Comment (alexdesiqueira) [inline skimage/filters/thresholding.py:242]: Could you please put this sentence in a `Notes` section?

File: skimage/filters/thresholding.py
Comment (alexdesiqueira) [inline skimage/filters/thresholding.py:285]: Same here. Could this sentence be in `Notes`?

File: skimage/filters/thresholding.py
Comment (alexdesiqueira) [inline skimage/filters/thresholding.py:355]: Same here: this sentence in `Notes` :)

File: skimage/filters/thresholding.py
Comment (alexdesiqueira) [inline skimage/filters/thresholding.py:430]: See previous :)

File: skimage/filters/thresholding.py
Comment (alexdesiqueira) [inline skimage/filters/thresholding.py:713]: Both sentences here in `Notes`, maybe?

**Review by alexdesiqueira — COMMENTED** [review]
Danke schön, @haesleinhuepf! I put some ideas around; I think it'd be nicer if we maintain the details on a `Notes` section in each doc, uncluttering the headers a little. How does that sound?
Could we also use `should` where we are using `must`, maybe? :slightly_smiling_face: 
Thanks again!

File: skimage/filters/thresholding.py
Comment (stefanv) [reply skimage/filters/thresholding.py:242]: These should be in parameter descriptions.  "Must" is a good word to use when, well, you *must* provide the argument for it to work :)

File: skimage/filters/thresholding.py
Comment (stefanv) [reply skimage/filters/thresholding.py:278]: Black can be told to leave quotes alone; that's my preference.

**Review by stefanv — APPROVED** [review]
@haesleinhuepf Thank you for your PR!

This is a good improvement; if folks feel strongly about any of the tweaks raised here, let's address them in a follow-up PR.

**Pull request merged**
